_posts/18-05-2022-a_story_unfolds.md

```diff
@@ -0,0 +1,223 @@
+
+What we get from separating the tree traversal from listing new branches is, again, quite subtle. First, we know that our traversal works for every structure that at least resembles a tree - and it is ensured by Haskell type system. Second, finding stop condition-related bugs is now easier, as we shifted the responsibility for terminating the recursion to functions that _don't have to be recursive at all!_ Imagine how easy testing becomes!
+
+The beauty of abstract constructs lies in their duality - they are incredibly rigid in one aspect, because their behavior is set in stone and virtually nothing can break it, and at the same time extremely flexible thanks to their generality. Once we create such a function, we can use it with confidence anywhere we wish (provided we can find a use case). The only downside is we need to spend more time to understand what is going on each time we encounter a new function - the more general it is, the harder to grasp it becomes. 
+
+My `treeBuilderM` is, however, far from the most beautiful function. It can get better, prettier and even more general. The forest is thicker and spreads further, with many paths left to explore.
+
+[^excuses]: I am by no means a Haskell expert, as I am still learning the language. Much of the code from this post could be written in a more elegant way, but I either don't know the proper constructs yet, or I am ignoring things like `ifM` (and even `$`) on purpose - to make the snippets more accessible to people from outside the Haskell world.
+
+[^axes]:
+    I am consciously ignoring the existence of symbolic links, along with the fact that a given file might not exist - for the sake of brevity.
+    And here is the implementation, based on `listDirectory` and `doesDirectoryExist` from `System.Directory` module, from [`directory`](https://hackage.haskell.org/package/directory) package:
+    ```haskell
+    data File = File { base :: Maybe FilePath, name :: FilePath }
+
+    fullPath :: File -> FilePath
+    fullPath (File (Just base) name) = base ++ "/" ++ name
+    fullPath (File Nothing name) = name
+
+    listDir :: File -> IO [File]
+    listDir file = do
+      let dirPath = fullPath file
+      contents <- listDirectory dirPath
+      return (map (File (Just dirPath)) contents)
+
+    isDir :: File -> IO Bool
+    isDir = doesDirectoryExist . fullPath
+    ```
+    `FilePath` is an alias for `String`, defined in `System.Directory`.
+
+    I decided to create the `File` type to avoid having to change the working directory, as the `listDirectory` function returns only the names of the files inside of a dir. It's not a proper implementation by any means, and it serves a purpose of simplifying the code a bit. It does also have a nasty side effect of making simple paths like `"."` into monstrosities like `File Nothing "."`.
```

Comment:
Also! will the code from the samples be present somewhere in the repo alongside the post? For the users to fiddle around

Reply:
Maybe even a link to some online haskell editor? if such a thing is available & reasonable

Reply:
Good point, will include that.